runtime/src/chplsys.c

```diff
@@ -399,23 +411,82 @@ uint64_t chpl_bytesPerLocale(void) {
 #endif
 }
 
+// Should return in some sense "currently available user memory"
+// This should return one of:
+//  * the amount of memory that can be allocated without causing swapping
+//  * the amount of memory that is currently unnused (i.e. free)
+uint64_t chpl_sys_availMemoryBytes(void) {
+#if defined(__APPLE__)
+
+  // On Mac OS X, we could get this value by calling host_statistics64
+  // but that's not really documented.
+  //
+  // So, we run vm_stat, which calls host_statistics, but
+  // which hopefully has a more stable interface.
+  //
+  // We used to use sysctlbyname("hw.usermem", &membytes, &len, NULL, 0)
+  // but that led to nonsense values being returned on some systems, anyway.
+
+  uint64_t pageSize = chpl_getSysPageSize();
+  FILE* f = popen("vm_stat", "r");
+  char buffer[256];
+  unsigned long pagesFree = 0;
+  int foundPagesFree = 0;
+
+  if (!f)
+    chpl_internal_error("could not run vm_stat in chpl_sys_availMemoryBytes");
+
+  while( fgets(buffer, sizeof(buffer), f) ) {
+    if (1 == sscanf(buffer, "Pages free: %lu", &pagesFree)) {
+      foundPagesFree = 1;
+    }
+  }
+
+  pclose(f);
+
+  if (!foundPagesFree)
+    chpl_internal_error("could not find Pages free vm_stat output "
+                        "in chpl_sys_availMemoryBytes");
+
+  return pageSize * pagesFree;
+
+#elif defined(__NetBSD__) || defined(__FreeBSD__)
+  // This used to run
+  //   sysctlbyname("hw.usermem64", &membytes, &len, NULL, 0)
+  // which returns the amount of non-kernel memory.
+  // It now gathers the amount of free memory to be more consistent
+  // with the Linux version.
+
+  uint64_t pageSize = chpl_getSysPageSize();
+  FILE* f = popen("vmstat -s", "r");
+  char buffer[256];
+  unsigned long pagesFree = 0;
+  int foundPagesFree = 0;
+
+  if (!f)
+    chpl_internal_error("could not run vm_stat in chpl_sys_availMemoryBytes");
+
+  while( fgets(buffer, sizeof(buffer), f) ) {
+    if (strstr(buffer, "pages free\n")) {
+      if (1 == sscanf(buffer, "%lu", &pagesFree))
+        foundPagesFree = 1;
+    }
+  }
+
+  pclose(f);
+
+  if (!foundPagesFree)
+    chpl_internal_error("could not find pages free vm_stat output "
+                        "in chpl_sys_availMemoryBytes");
+
+  return pageSize * pagesFree;
 
-size_t chpl_bytesAvailOnThisLocale(void) {
-#if defined __APPLE__
-  int membytes;
-  size_t len = sizeof(membytes);
-  if (sysctlbyname("hw.usermem", &membytes, &len, NULL, 0)) 
-    chpl_internal_error("query of physical memory failed");
-  return (size_t) membytes;
-#elif defined __NetBSD__
-  int64_t membytes;
-  size_t len = sizeof(membytes);
-  if (sysctlbyname("hw.usermem64", &membytes, &len, NULL, 0))
-    chpl_internal_error("query of hw.usermem64 failed");
-  return (size_t) membytes;
 #elif defined(__linux__)
   struct sysinfo s;
 
+  // MPF: I think this should return MemAvailable not MemFree
```

Comment:
If the change described in this comment would result in the value returned for Linux being more similar to what we return on other platforms, then you should go ahead and do it now, unless there's a reason to put it off that I'm not aware of.

Reply:
This commit makes the other platforms be in line with linux - namely they all currently return the amount of free memory. I don't think that's the right answer long-term (because of the page cache) but I think we need to first resolve issue #8199 and understand a little bit the performance impact of changing this number (since we're not using this value on Cray systems currently anyway to set the GASNet segment size - which is all we do with this function). E.g. if it turns out not to matter for benchmarks, we might remove this function entirely, or tolerate its continued inaccuracies.